### ISSUE TITLE
Help on using logsumexp during SGD training

### ISSUE THREAD

**Issue by duyvuleo** [clab/dynet#481]
Hi everyone,

In my problem, I want to implement the exponentiated gradient update as in Kivinen & Warmuth (1997).
(Kivinen, Jyrki and Warmuth, Manfred K. Exponentiated gradient versus gradient descent for linear predictors. Information and Computation, 132(1):1–63, 1997)

I would like to utilise the current implementation of logsumexp during SGD training in which I apply to computation of parameters/gradients. 
The update formula of EG is defined as y_{t+1} := 1/Z * y_t * exp(-eta * dL(y_t)) where Z = \sum_i y_{t,i} * exp(-eta * dL(y_{t,i})).
To avoid the underflow/overflow issues, I take the log of y_{t+1}, resulting in: log(y_{t+1}) = log(y_t) - eta * dL(y_t) - logZ.
Here, logZ refers to logsumexp.

Assume that I have new SGD trainer denoted as EGTrainer having the update_rule_dev function implemented as follows:
```
template <class MyDevice>
void EGTrainer::update_rule_dev(const MyDevice & dev, real scale, real gscale, const std::vector<Tensor*> & ts) {
  //------------------------------------------------------------------
  // Add momentum
  ts[2]->tvec().device(*dev.edevice) = ts[2]->tvec() * momentum - ts[1]->tvec() * (eta * scale * gscale);
  //------------------------------------------------------------------

  //------------------------------------------------------------------
  // METHOD 2: advanced implementation (supposedly faster)
  ts[0]->tvec().device(*dev.edevice) = ts[0]->tvec().log() + ts[2]->tvec() / model->weight_decay.current_weight_decay();// with momentum

  float* aux_mem = 0;// how to set this?

  Tensor z(Dim({1}, 1), (float*)aux_mem, ts[0]->device, DeviceMempool::FXS);
  Tensor m(Dim({1}, 1), (float*)aux_mem + z.d.size(), ts[0]->device, DeviceMempool::FXS);

  logsumexp(dev, *ts[0], m, z);

  ts[0]->tvec().device(*dev.edevice) = (ts[0]->tvec() - as_scalar(z)).exp();
  //------------------------------------------------------------------
}
```

I am confused with the followings:
1) Are the dims of z and m correct?
2) How to set aux_mem for initialising the tensors z and m?
3) is DeviceMempool::FXS correct?

I appreciate if someone can help on this!

Thank you very much!

--
Vu

**Comment by neubig** [comment]
Hi Vu,

This is interesting! Please feel free to contribute the EGTrainer back if you get it working properly. Anyway, answers to the questions:

1. Yes, I think so. Incidentally you don't need to write `Dim({1}, 1)`, just `Dim({1})` or `{1}` is fine.
2. Setting `nullptr` in the instructor, and then calling `default_device->allocate_tensor(DeviceMempool::PS, z);` should work.
3. `DeviceMempool::PS` is the parameter pool.

**Comment by duyvuleo** [comment]
Thanks Graham. It works now. 
I am happy to contribute the following code:

```
// Written by Cong Duy Vu Hoang
/**
 * \ingroup optimizers
 * 
 * \brief Exponentiated gradient optimizer with momentum and cyclical learning rate
 * \details FIXME
 *  
 * Reference : FIXME
 *   
*/
struct EGTrainer : public Trainer {
  explicit EGTrainer(Model& mod, real e0 = 0.1, real mom = 0.9, real edecay = 0.0)
    : Trainer(mod, e0, edecay), momentum(mom), isCyclical(false) {
    zeg.d = meg.d = {1};
    zeg.device = meg.device = default_device;
    default_device->allocate_tensor(DeviceMempool::PS, zeg);
    default_device->allocate_tensor(DeviceMempool::PS, meg);
  }

//-----------------------------------------------------------------------------------------
  void enableCyclicalLR(float _e0_min = 0.01, float _e0_max = 0.1, float _step_size = 2000, float _gamma = 0.0){
    isCyclical = true;
    e_min = _e0_min;
    e_max = _e0_max;
    step_size = _step_size;
    gamma = _gamma;
    it = 0;
  }

  virtual void update(real scale = 1.0) { 
    Trainer::update(scale); 
    if (isCyclical) cyclic_update_eta();
  }
//-----------------------------------------------------------------------------------------

 protected:
  DYNET_TRAINER_DEFINE_DEV_IMPL()
  virtual void alloc_impl() override;

//-----------------------------------------------------------------------------------------
  real momentum;// with momentum
  std::vector<ShadowParameters> hp; // (previous) history of parameters
  std::vector<ShadowLookupParameters> hlp;// unused at the moment
//-----------------------------------------------------------------------------------------

//-----------------------------------------------------------------------------------------
  void cyclic_update_eta() {
    float cycle = std::floor(1 + ((float) it)  / (2 * step_size));
    float x = std::abs( ((float) it) / step_size - 2 * cycle + 1);
    //std::cerr << "cycle=" << cycle << "; " << "x=" << x << std::endl;
    eta = e_min + ((1 - x) > 0 ? (e_max - e_min) * (1 - x) * std::pow(gamma, it) : 0);
    it++;
  }

  float e_min = 0;
  float e_max = 0;
  float step_size = 0;
  float gamma = 0;
  unsigned it = 0;
  bool isCyclical;
//-----------------------------------------------------------------------------------------

//-----------------------------------------------------------------------------------------
// temporary tensors for EG calculation
  Tensor zeg, meg;
//-----------------------------------------------------------------------------------------

 private:
  EGTrainer() {}
  DYNET_SERIALIZE_DECLARE()
};
BOOST_CLASS_EXPORT_KEY(dynet::EGTrainer)

// --- EGTrainer
template <class MyDevice>
void EGTrainer::update_rule_dev(const MyDevice & dev, real scale, real gscale, const std::vector<Tensor*> & ts) {
  //------------------------------------------------------------------
  // Add white Gaussian noise
  //real stddev = scale;// FIXME: should be an *annealing* factor
  //TensorTools::RandomizeNormal(0, stddev, *ts[2]);
  //ts[1]->tvec().device(*dev.edevice) += ts[2]->tvec();

  //------------------------------------------------------------------
  // Add momentum
  ts[2]->tvec().device(*dev.edevice) = ts[2]->tvec() * momentum - ts[1]->tvec() * (eta * scale * gscale);

/*
  //------------------------------------------------------------------
  // METHOD 1: naive implementation (slow?)
  std::vector<real> v_params = dynet::as_vector(*ts[0]);// FIXME: not a smart way but it works on both GPU and CPU!
  std::vector<real> v_grads = dynet::as_vector(*ts[2]);// gradients with momentum
  Eigen::TensorMap<Eigen::Tensor<real,1>> t_params(&v_params[0], v_params.size());
  Eigen::TensorMap<Eigen::Tensor<real,1>> t_grads(&v_grads[0], v_grads.size());
  Eigen::Tensor<real,1> v = t_params.log() + t_grads / model->weight_decay.current_weight_decay();// with momentum
  
  Eigen::Tensor<float,0> m = v.maximum();// max
  Eigen::Tensor<float,0> z = m.coeff() + (v - m.coeff()).exp().sum().log();// Z
  Eigen::Tensor<float,1> u = (v - z.coeff()).exp();// result
  TensorTools::set_elements(*ts[0], u.data(), u.dimensions()[0]);// for both GPU and CPU
  //------------------------------------------------------------------
*/

  //------------------------------------------------------------------
  // METHOD 2: advanced implementation (supposedly faster?)
  ts[0]->tvec().device(*dev.edevice) = ts[0]->tvec().log() + ts[2]->tvec() / model->weight_decay.current_weight_decay();// with momentum

  logsumexp(dev, *ts[0], *ts[3], *ts[4]);// z refers to logZ
  //cerr << "max_element=" << as_scalar(m) << endl;
  //cerr << "logZ=" << as_scalar(z) << endl;
    
  ts[0]->tvec().device(*dev.edevice) = (ts[0]->tvec() - as_scalar(*ts[4])).exp();// FIXME: other way(s) of not using as_scalar(z)?
  //------------------------------------------------------------------
}
DYNET_TRAINER_INST_DEV_IMPL(EGTrainer)

#ifndef __CUDACC__
void EGTrainer::update_params(real scale, real gscale, size_t idx) {
  auto & p = model->parameters_list()[idx];
  update_rule(scale, gscale, {&p->values, &p->g, &hp[idx].h, &meg, &zeg});
}
void EGTrainer::update_lookup_params(real scale, real gscale, size_t idx, size_t lidx) {
  auto & p = model->lookup_parameters_list()[idx];
  update_rule(scale, gscale, {&p->values[lidx], &p->grads[lidx], &hlp[idx].h[lidx], &meg, &zeg});
}
void EGTrainer::update_lookup_params(real scale, real gscale, size_t idx) {
  auto & p = model->lookup_parameters_list()[idx];
  update_rule(scale, gscale, {&p->all_grads, &p->all_grads, &hlp[idx].all_h, &meg, &zeg});
}
void EGTrainer::alloc_impl() {
  hp = allocate_shadow_parameters(*model);
  hlp = allocate_shadow_lookup_parameters(*model); 
}
#endif

DYNET_SERIALIZE_COMMIT(EGTrainer, DYNET_SERIALIZE_DERIVED_DEFINE(Trainer, momentum, hp, hlp))
DYNET_SERIALIZE_IMPL(EGTrainer)

```

**Comment by neubig** [comment]
If you don't mind sending it as a pull request, that would be great! Of course, I can integrate it directly as well if you'd prefer.

**Comment by duyvuleo** [comment]
Sure. I will do a pull request shortly.

**Comment by duyvuleo** [comment]
Refers to #545 pull request.
